### ISSUE TITLE
`inject` is not working as a reactive function in version 4.1.8 

### ISSUE THREAD

**Issue by robinspark** [mobxjs/mobx-react#283]
According to [this resolved issue](https://github.com/mobxjs/mobx-react/issues/111), primitives should be tracked inside @inject blocks, and be 'reactive', as of mobx-react version 4.0.0

I've created an example where this doesn't work correctly using version 4.1.8.

It doesn't work where a primitive is destructured inside the @inject block before being passed in as props.

In the following examples, selecting a value in the dropdown calls:
  setFormFilter(event.target.value); //update the form id in the store.
  renderResponse(); //calls a function in the parent which should have access to the updated formFilter value from the store.

Passing the whole state object from store works.
[https://codesandbox.io/s/1r7YM60lR](https://codesandbox.io/s/1r7YM60lR)
componentWillReact is called immediately after setFormFilter(), and before renderResponse().

Destructure the primitive formFilter inside the stores state doesn't work.
[https://codesandbox.io/s/9QjVnKq1Z](https://codesandbox.io/s/9QjVnKq1Z)
componentWillReact is called after both the setFormFilter(), and renderResponse() calls.

**Comment by mweststrate** [comment]
Can you clarifiy your example a bit? Some remarks:

`  const { setFormFilter, state: responseState } = store;`. store.state doesn't seem to contain a `responseState` field? So I am not sure what is supposed to happen there

```
extendObservable(this.state, {
      formFilter: formFilter,
    });
``` 
doesn't do anything special, you cansimply write `this.state.formFilter = formFilter`

I would appreciate it if the example is simplified; it is currently pretto to hard to follow, and it is a bit unclear to me what you are trying to acieve in the first place :wink: 

For example what is a callback like `renderResponse` doing? Your not supposed to trigger rendering yourself in the first place

**Comment by robinspark** [comment]
(renderResponse() is just an method in the parent Component. It doesn't force re-render.)
The issue is that when a state property is deconstructed in the inject block, 

For the 2 examples above, open the browser console, change the dropdown value and compare the output. Depending on whether you pass the whole state vs Destructuring the state, the store values do not update at the same point in time.

The first example (pass whole state): store.formFilter is 1 //correct
The second example (pass destructured state): store.formFilter: -1 //incorrect.

**Comment by mweststrate** [comment]
As said, your store is strangely set up, as `state` is initially not an observable property, but you make it an observable property only when the `setFormFilter` function is called for the first time. Changed that and now both examples provide the same output (although the first has the same erronous setup as the second one had). https://codesandbox.io/s/jE7RrZy

If you have follow up questions, feel free to reopen, please mention what you expect the output to be, and why.

**Comment by robinspark** [comment]
> state is initially not an observable property, but you make it an observable property only when the setFormFilter function is called for the first time.

Does my original code (below) not initialise state as an observable?
```
state = observable({
    formFilter: -1,
});
```

**Comment by mweststrate** [comment]
@robinspark your first example makes `{ formFilter }` object observable, but not `state`. So changing the .formFilter attribute would be picked up, but a reassignment to the `state` property doesn't. Unlike the second approach which makes both observable. (`@observable` is like `extendObservable(this, { state: { formFilter }})`
